File: app/src/main/java/com/example/rickmortymvvm/services/ListCharacterAdapter.java
package com.example.rickmortymvvm.services;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.rickmortymvvm.R;
import com.example.rickmortymvvm.intrefaces.OnClickCharacter;

import java.util.ArrayList;
import java.util.List;

import com.example.rickmortymvvm.models.Character;
import com.example.rickmortymvvm.view.PresentationCharacterListActivity;

public class ListCharacterAdapter extends RecyclerView.Adapter<ListCharacterAdapter.ViewHolder> {

    private List<Character> list = new ArrayList<>();
    private PresentationCharacterListActivity presentationCharacterListActivity;


    public ListCharacterAdapter(PresentationCharacterListActivity presentationCharacterListActivity) {
        this.presentationCharacterListActivity = presentationCharacterListActivity;
    }

    public void setListAdapter(List<Character> listCharacter) {
        list =listCharacter;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ListCharacterAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.model_main_activity, parent, false);
        return new ViewHolder(view, presentationCharacterListActivity);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(list.get(position));
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageView;
        private final TextView nameView;
        private final TextView statusView;
        private Character character;

        public ViewHolder(@NonNull View itemView, OnClickCharacter clickCharacter) {
            super(itemView);
            imageView = itemView.findViewById(R.id.image);
            nameView = itemView.findViewById(R.id.name);
            statusView = itemView.findViewById(R.id.status);

        }


        public void bind(Character character) {
            this.character = character;
            Glide.with(imageView)
                    .load(character.getImage())
                    .into(imageView);
            nameView.setText(character.getName());
            statusView.setText(character.getStatus());
            if (character.getStatus().contains("Dead")) {
                int color = ContextCompat.getColor(statusView.getContext(), R.color.colorRed);
                statusView.setTextColor(color);
            } else if (character.getStatus().contains("Alive")) {
                int color = ContextCompat.getColor(statusView.getContext(), R.color.colorgreem);
                statusView.setTextColor(color);
            } else {
                int color = ContextCompat.getColor(statusView.getContext(), R.color.colorBlck);
                statusView.setTextColor(color);
            }
        }

    }
}

File: app/src/main/java/com/example/rickmortymvvm/view/ListCharacterAdapter.kt
package com.example.rickmortymvvm.app.view

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.rickmortymvvm.R
import com.example.rickmortymvvm.databinding.ModelMainActivityBinding
import com.example.rickmortymvvm.presentation.models.Character
import com.example.rickmortymvvm.app.util.adapter.DiffUtilGeneric

class ListCharacterAdapter(private val presentationCharacterListActivity: PresentationCharacterListActivity) :
    RecyclerView.Adapter<ListCharacterAdapter.ViewHolder>() {
    private val list: MutableList<Character> = ArrayList()

    fun setListAdapter(newList: List<Character>) {
        val diffResult = DiffUtil.calculateDiff(DiffUtilGeneric(list, newList))
        list.clear()
        list.addAll(newList)

        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ModelMainActivityBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return ViewHolder(binding, presentationCharacterListActivity)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(list[position])
    }

    override fun getItemCount(): Int {
        return list.size
    }

    class ViewHolder(binding: ModelMainActivityBinding, clickCharacter: OnClickCharacter) :
        RecyclerView.ViewHolder(binding.root) {

        private val imageView = binding.image
        private val nameView = binding.name
        private val statusView = binding.status
        private lateinit var character: Character

        fun bind(character: Character) {
            this.character = character
            Glide.with(imageView)
                .load(character.image)
                .into(imageView)
            nameView.text = character.name
            statusView.text = character.status

            if (character.status == "Dead") {
                val color = ContextCompat.getColor(statusView.context, R.color.colorRed)
                statusView.setTextColor(color)
            } else if (character.status == "Alive") {
                val color = ContextCompat.getColor(statusView.context, R.color.colorgreem)
                statusView.setTextColor(color)
            } else {
                val color = ContextCompat.getColor(statusView.context, R.color.colorBlck)
                statusView.setTextColor(color)
            }
        }

        init {
            binding.root.setOnClickListener {
                clickCharacter.onClickCharacter(character)
            }
        }
    }
}
